lib/Basics/StringUtils.cpp

```diff
@@ -1300,6 +1303,84 @@ std::string urlEncode(char const* src, size_t const len) {
 
   return result;
 }
+    
+std::string encodeURIComponent(std::string const& str) {
+  return encodeURIComponent(str.c_str(), str.size());
+}
+    
+std::string encodeURIComponent(char const* src, size_t const len){
+  char const* end = src + len;
+    
+  if (len >= (SIZE_MAX - 1) / 3) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+  }
+    
+  std::string result;
+  result.reserve(3 * len);
+    
+  for(; src < end; ++src){
+    if (*src == '-' || *src == '_' || *src == '.' || *src == '!' || *src == '~' || *src == '*' || *src == '(' || *src == ')' || *src == '\''|| (*src >= 'a' && *src <= 'z') || (*src >= 'A' && *src <= 'Z') || (*src >= '0' && *src <= '9')) {
+      result.push_back(*src);
+    }
+    else{
+      result.push_back('%');
+      result = result + encodeHex(src, 1);
+    }
+  }
+    
+  return result;
+}
+    
+std::string soundex(std::string const& str){
+    if (str == ""){
+        return "0000";
+    }
+    return soundex(str.c_str(), str.size());
+}
+    
+char soundexCode(const char& c){
+    switch (c) {
+        case 'b': case 'f': case 'p': case 'v':
+            return '1';
+        case 'c': case 'g': case 'j': case 'k': case 'q': case 's': case 'x': case 'z':
+            return '2';
+        case 'd': case 't':
+            return '3';
+        case 'l':
+            return '4';
+        case 'm': case 'n':
+            return '5';
+        case 'r':
+            return '6';
+        default:
+            return '\0';
+    }
+}
+    
+std::string soundex(char const* src, size_t const len){
+    char const* end = src + len;
+    std::string result;
+    char previousCode = '\0';
+    
+    result.push_back(::toupper(*src));
+    src++;
+    
+    while (src < end) {
+        char currentCode = soundexCode(*src);
+        if (currentCode != '\0' && currentCode != previousCode && result.length() < 4) {
```

Comment:
The check for `result.length() < 4` should go after the `result.push_back(currentCode)` line.
Because only the `push_back` will change the length of `result`, so the check for length 4 should be directly after the `push_back`.